File: lib/vector.js
/**
 * Declaration of Vector(x, y), one of the basic structures needed
 */
var Vector = function(x, y) {
    this.x = x; // coordinate by x-axis
    this.y = y; // coordinate by y-axis
};

/**
 * Build Vector by two Points
 */
var buildVector = function(begin, end) {
    // we should substract x and y coordiantes of Point begin
    // respectively from x and y coordiantes of Point end
    return new Vector(end.x - begin.x, end.y - begin.y);
};

/**
 * Implementation of crossproduct for two vectors
 */
var crossProduct = function (v1, v2) {
    return v1.x * v2.y - v1.y * v2.x;
};

/**
 * Implementation of dot product for two vectors
 */
var dotProduct = function (v1, v2) {
    return v1.x * v2.x + v1.y * v2.y;
};

/**
 * Rotates vector for a current angle
 */
var rotateVector = function (v, angle) {
    // while rotating Vector, we change the coordinates according to the rotation matrix
    return new Vector(v.x * Math.cos(angle) - v.y * Math.sin(angle), v.x * Math.sin(angle) + v.y * Math.cos(angle));
};

File: slines_py/lib/vector.py
'''
Decloration of Vector, one of the basic structures needed
'''

import math

class Vector:
    def __init__(self, _arg1, _arg2):
        if type(_arg1) == int:
            self.x = _arg1
            self.y = _arg2
        else:
            self.x = _arg2.x - _arg1.x
            self.y = _arg2.y - _arg1.y

def crossProduct(_v1, _v2):
    return _v1.x * _v2.y - _v1.y * _v2.x

def dotProduct(_v1, _v2):
    return _v1.x * _v2.x + _v1.y * _v2.y


'''
Rotates vector for a current angle
'''
def rotateVecor(_v, _angle):
    return Vector(_v.x * math.cos(_angle) - _v.y * math.sin(_angle), _v.x * math.sin(_angle) + _v.y * math.cos(_angle))
